Improve message for misplaced using

som-snytt opened the pull request:
Offer extra feedback when soft keyword `using` is found in identifier position, because it is a common beginner mistake.

```
def f(i: Int, using j: Int) = i + j
```

Skipping to the intended `j` may also reduce subsequent "not found" errors.

In Vulpix, show actual errors when no errors are marked. It is unlikely that this is "TMI" for the developer. Automatic `-rewrite` is left for another day.

Fixes #15614

odersky commented on compiler/src/dotty/tools/dotc/parsing/Parsers.scala:
```diff
@@ -3095,8 +3095,14 @@ object Parsers {
-          acceptColon()
+          val name =
+            val name0 = ident()
+            if name0 == nme.using && !in.isColon then
```
I think the same logic should be applied `erased`. Like using, it is a soft modifier that can only appear at the start of a parameter clause.

som-snytt replied:
What's `erased` again? Oh, yeah! Refresher. Thanks for the quick review.

som-snytt replied:
Apparently erase syntax was updated last year.

odersky reviewed (changes requested):
Otherwise LGTM

odersky commented on compiler/src/dotty/tools/dotc/parsing/Tokens.scala:
```diff
@@ -179,6 +179,8 @@ object Tokens extends TokensCommon {
   inline val EXPORT = 64;           enter(EXPORT, "export")
   inline val MACRO = 65;            enter(MACRO, "macro") // TODO: remove
   inline val END = 66;              enter(END, "end")
+  inline val USING = 67;            enter(USING, "using")
```
That would make it a hard modifier.

som-snytt replied:
It's not actually tokenized, but the value is used for the error message mechanism. Oh, the enter part. Cut, paste, unpaste, repeat.

som-snytt replied:
That just gives it a `tokenString`, as opposed to registering a `keyword`.

odersky replied:
I think that's way too confusing.

som-snytt replied:
The message can just take the name instead of a token.

som-snytt commented on compiler/src/dotty/tools/dotc/reporting/messages.scala:
```diff
@@ -1143,6 +1144,18 @@ import transform.SymUtils._
+  class ExpectedTokenButFoundSoftToken(expected: Token, found: Token, soft: Name)(using Context) extends SyntaxMsg(ExpectedTokenButFoundID):
+    def msg =
+      val expectedText =
+        if (Tokens.isIdentifier(expected)) "an identifier"
```
cut/paste parens

som-snytt commented:
Thanks again for the quick review. A little bit of experienced feedback goes a long way.

Updated without tokens for using, erased. Worth noting that `using` is enshrined as a standard name, but tracking down usages is still tricky. Similarly, I wondered why `then` isn't rewritten. What is handled where? Every special case increases complexity.

Modified the verbose advice to say an error makes a symbol unavailable. That happens frequently during a refactor, where fixing one site results in further errors in the next run.

som-snytt commented:
@odersky Please approve based on Bjorn's approval.

bjornregnell commented:
I may be missing some context here, but why closed instead of merged? @som-snytt

som-snytt commented:
rebased, squashed, removed some words during ugly merge.

som-snytt commented:
@bjornregnell Could you re-review and perhaps get official sanction?

The previous effort included heavier edit of the explain text, which I omit here because of the heavy merge.

The previous effort included handling `erased`, which is no longer needed.

The only trick in parser is error recovery in skipping bad token to the presumably intended ident.

I'll also update the introductory comment above.

bjornregnell commented:
LGTM
Thanks for nice work in increasing quality of life on planet Earth @som-snytt 
I think this is ready to be merged @odersky - who should do that?

som-snytt commented:
Reopening to reconsider this forgotten bit.

som-snytt commented:
I'm not sure it was ever relevant, but I have refreshed the PR.

Gedochao commented:
(looks good to me, but let's do a fresh review; I'll make sure we wouldn't let it get stale again)

SolalPirelli reviewed (approved):
Clearly better than the current state. I think "in certain contexts" as an explanation is not great, a newcomer won't know what those contexts are, and we could explicitly suggest splitting the parameter list into two. But that's probably for another day.
